ecpp/license_POST.ecpp

```diff
@@ -142,128 +144,178 @@ bool database_ready;
+        }
+        log_debug ("Starting db service will refresh tntnet deadman timer every %u msec", wait_cycle);
+
+        for (;;) { // repeat_start_db_services:
+            // May finish before "wait_cycle" elapses:
+            auto rv_svc = proc.wait(wait_cycle);
+            uint64_t tme_finish = uint64_t(::time (NULL));
+
+            // Success? Fail? Keep waiting?..
+            if (!rv_svc && "timeout" == rv_svc.error()) {
+                // ...keep waiting, script is not finished:
+                log_info ("Starting db service takes longer than expected on this system, waiting more, timestamp=%" PRIu64 " (duration=%" PRIu64 ")" ,
+                    tme_finish, (tme_finish - tme_start));
+                request.touch();
+                continue; // goto repeat_start_db_services;
             }
-            // Negative return means killed by signal, typically by our
-            // subprocess timeout handling - fall through to try using DB
 
-            {
-                std::string proc_out1, proc_err1;
-                int lastResult = 0;
-                for (int i = 0; i < 10; ++i) {
-                    fty::Process::Arguments proc_cmd = {"systemctl", "list-ipm-units", "--active", "fty-db-engine"};
-                    // This should get our "systemctl" wrapper via PATH
+            // ...success or fail, script is finished
+            // Read the outputs:
+            proc_out = proc.readAllStandardOutput();
+            proc_err = proc.readAllStandardError();
+            // Terminate child (if any) to avoid leaks:
+            proc.kill();
+
+            if (rv_svc) {
+                log_info ("Starting db service completed, timestamp=%" PRIu64 " (duration=%" PRIu64 "), exit-code=%i ..." ,
+                    tme_finish, (tme_finish - tme_start), *rv_svc);
+            } else {
+                log_info ("Starting db service completed, timestamp=%" PRIu64 " (duration=%" PRIu64 "), exit-code=%s ..." ,
+                    tme_finish, (tme_finish - tme_start), rv_svc.error().c_str());
+            }
 
-                    auto rv = fty::Process::run("sudo", proc_cmd, proc_out1, proc_err1);
+            // non-zero exit code or proc class exception:
+            if (!rv_svc || *rv_svc != 0) {
+                log_error ("Starting of start-db-services have failed. Consult system logs");
+                log_error ("%s failed with error %s.\n===== STDOUT: =====\n%s\n===== STDERR: =====\n%s\n=====",
+                    "/usr/libexec/fty/start-db-services", rv_svc.error().c_str(), proc_out.c_str (), proc_err.c_str ());
 
-                    if (!rv) {
-                        log_error(rv.error().c_str());
-                        log_error_audit ("Request CREATE license FAILED");
-                        http_die ("internal-error", TRANSLATE_ME ("Database software service is not running").c_str());
+                if (rv_svc.error() != "timeout") {
```

Comment:
Could be problem.
you allow to go into this scope by *rv_svc != 0... means this variable has value, not error. But in this line you check error of possible value.

Reply:
Sorry, got no opinion about that. Just indented original code as it was. Just looking around, apparently the `Expected<int>` which that `auto rv_svc` is, works that way. Similarly, no idea why `auto` instead of specific types with expected behaviors.

Reply:
Thanks for the catch, confirmed as problematic by killing the script (via terminal) while it was being processed. Fix pending, but workstation got no internet to post it at the moment.